src/domain/common/entities/safe-signature.ts

```diff
@@ -0,0 +1,127 @@
+import { secp256k1 } from '@noble/curves/secp256k1';
+import { memoize } from 'lodash';
+import {
+  checksumAddress,
+  encodeAbiParameters,
+  hashMessage,
+  hexToBigInt,
+  parseAbiParameters,
+} from 'viem';
+import { publicKeyToAddress } from 'viem/utils';
+import { SignatureType } from '@/domain/common/entities/signature-type.entity';
+
+export class SafeSignature {
+  public signature: `0x${string}`;
+  public hash: `0x${string}`;
+
+  constructor(args: { signature: `0x${string}`; hash: `0x${string}` }) {
+    if (args.signature.length !== 132) {
+      throw new Error('Invalid signature length');
+    }
+
+    this.signature = args.signature;
+    this.hash = args.hash;
+  }
+
+  get r(): `0x${string}` {
+    return `0x${this.signature.slice(2, 66)}`;
+  }
+
+  get s(): `0x${string}` {
+    return `0x${this.signature.slice(66, 130)}`;
+  }
+
+  get v(): number {
+    return parseInt(this.signature.slice(-2), 16);
+  }
+
+  get signatureType(): SignatureType {
+    if (this.v === 0) {
+      return SignatureType.ContractSignature;
+    }
+    if (this.v === 1) {
+      return SignatureType.ApprovedHash;
+    }
+    if (this.v > 30) {
+      return SignatureType.EthSign;
+    }
+    return SignatureType.Eoa;
+  }
+
+  get owner(): `0x${string}` {
+    return this._owner();
+  }
+
+  private _owner = memoize(
+    () => {
+      try {
+        switch (this.signatureType) {
+          case SignatureType.ContractSignature:
+          case SignatureType.ApprovedHash: {
+            return uint256ToAddress(this.r);
+          }
+          case SignatureType.EthSign: {
```

Comment:
Nit: should we document _why_ is it needed to subtract 4 to the `v` value when `SignatureType` is `EthSign`?